mcc/mcc.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MCC 1"
.TH MCC 1 "2012-08-11" " " " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
tcc \- Tiny C Compiler
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
usage: tcc [options] [\fIinfile1\fR \fIinfile2\fR...] [\fB\-run\fR \fIinfile\fR \fIargs\fR...]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\s-1TCC\s0 options are a very much like gcc options. The main difference is that \s-1TCC\s0
can also execute directly the resulting program and give it runtime
arguments.
.PP
Here are some examples to understand the logic:
.ie n .IP """\f(CBtcc \-run a.c\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-run a.c\f(CW\fR" 4
.IX Item "tcc -run a.c"
Compile \fIa.c\fR and execute it directly
.ie n .IP """\f(CBtcc \-run a.c arg1\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-run a.c arg1\f(CW\fR" 4
.IX Item "tcc -run a.c arg1"
Compile a.c and execute it directly. arg1 is given as first argument to
the \f(CW\*(C`main()\*(C'\fR of a.c.
.ie n .IP """\f(CBtcc a.c \-run b.c arg1\f(CW""" 4
.el .IP "\f(CW\f(CBtcc a.c \-run b.c arg1\f(CW\fR" 4
.IX Item "tcc a.c -run b.c arg1"
Compile \fIa.c\fR and \fIb.c\fR, link them together and execute them. arg1 is given
as first argument to the \f(CW\*(C`main()\*(C'\fR of the resulting program.
.ie n .IP """\f(CBtcc \-o myprog a.c b.c\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-o myprog a.c b.c\f(CW\fR" 4
.IX Item "tcc -o myprog a.c b.c"
Compile \fIa.c\fR and \fIb.c\fR, link them and generate the executable \fImyprog\fR.
.ie n .IP """\f(CBtcc \-o myprog a.o b.o\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-o myprog a.o b.o\f(CW\fR" 4
.IX Item "tcc -o myprog a.o b.o"
link \fIa.o\fR and \fIb.o\fR together and generate the executable \fImyprog\fR.
.ie n .IP """\f(CBtcc \-c a.c\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-c a.c\f(CW\fR" 4
.IX Item "tcc -c a.c"
Compile \fIa.c\fR and generate object file \fIa.o\fR.
.ie n .IP """\f(CBtcc \-c asmfile.S\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-c asmfile.S\f(CW\fR" 4
.IX Item "tcc -c asmfile.S"
Preprocess with C preprocess and assemble \fIasmfile.S\fR and generate
object file \fIasmfile.o\fR.
.ie n .IP """\f(CBtcc \-c asmfile.s\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-c asmfile.s\f(CW\fR" 4
.IX Item "tcc -c asmfile.s"
Assemble (but not preprocess) \fIasmfile.s\fR and generate object file
\&\fIasmfile.o\fR.
.ie n .IP """\f(CBtcc \-r \-o ab.o a.c b.c\f(CW""" 4
.el .IP "\f(CW\f(CBtcc \-r \-o ab.o a.c b.c\f(CW\fR" 4
.IX Item "tcc -r -o ab.o a.c b.c"
Compile \fIa.c\fR and \fIb.c\fR, link them together and generate the object file \fIab.o\fR.
.PP
Scripting:
.PP
\&\s-1TCC\s0 can be invoked from \fIscripts\fR, just as shell scripts. You just
need to add \f(CW\*(C`#!/usr/local/bin/tcc \-run\*(C'\fR at the start of your C source:
.PP
.Vb 2
\&        #!/usr/local/bin/tcc \-run
\&        #include <stdio.h>
\&        
\&        int main() 
\&        {
\&            printf("Hello World\en");
\&            return 0;
\&        }
.Ve
.PP
\&\s-1TCC\s0 can read C source code from \fIstandard input\fR when \fB\-\fR is used in 
place of \fBinfile\fR. Example:
.PP
.Vb 1
\&        echo \*(Aqmain(){puts("hello");}\*(Aq | tcc \-run \-
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-v\fR" 4
.IX Item "-v"
Display current \s-1TCC\s0 version, increase verbosity.
.IP "\fB\-c\fR" 4
.IX Item "-c"
Generate an object file (\fB\-o\fR option must also be given).
.IP "\fB\-o outfile\fR" 4
.IX Item "-o outfile"
Put object file, executable, or dll into output file \fIoutfile\fR.
.IP "\fB\-Bdir\fR" 4
.IX Item "-Bdir"
Set the path where the tcc internal libraries can be found (default is
\&\fIPREFIX/lib/tcc\fR).
.IP "\fB\-bench\fR" 4
.IX Item "-bench"
Output compilation statistics.
.IP "\fB\-run source [args...]\fR" 4
.IX Item "-run source [args...]"
Compile file \fIsource\fR and run it with the command line arguments
\&\fIargs\fR. In order to be able to give more than one argument to a
script, several \s-1TCC\s0 options can be given \fIafter\fR the
\&\fB\-run\fR option, separated by spaces. Example:
.Sp
.Vb 1
\&        tcc "\-run \-L/usr/X11R6/lib \-lX11" ex4.c
.Ve
.Sp
In a script, it gives the following header:
.Sp
.Vb 6
\&        #!/usr/local/bin/tcc \-run \-L/usr/X11R6/lib \-lX11
\&        #include <stdlib.h>
\&        int main(int argc, char **argv)
\&        {
\&            ...
\&        }
.Ve
.PP
Preprocessor options:
.IP "\fB\-Idir\fR" 4
.IX Item "-Idir"
Specify an additional include path. Include paths are searched in the
order they are specified.
.Sp
System include paths are always searched after. The default system
include paths are: \fI/usr/local/include\fR, \fI/usr/include\fR
and \fIPREFIX/lib/tcc/include\fR. (\fI\s-1PREFIX\s0\fR is usually
\&\fI/usr\fR or \fI/usr/local\fR).
.IP "\fB\-Dsym[=val]\fR" 4
.IX Item "-Dsym[=val]"
Define preprocessor symbol \fBsym\fR to
val. If val is not present, its value is \fB1\fR. Function-like macros can
also be defined: \fB\-DF(a)=a+1\fR
.IP "\fB\-Usym\fR" 4
.IX Item "-Usym"
Undefine preprocessor symbol \fBsym\fR.
.PP
Compilation flags:
.PP
Note: each of the following warning options has a negative form beginning with
\&\fB\-fno\-\fR.
.IP "\fB\-funsigned\-char\fR" 4
.IX Item "-funsigned-char"
Let the \f(CW\*(C`char\*(C'\fR type be unsigned.
.IP "\fB\-fsigned\-char\fR" 4
.IX Item "-fsigned-char"
Let the \f(CW\*(C`char\*(C'\fR type be signed.
.IP "\fB\-fno\-common\fR" 4
.IX Item "-fno-common"
Do not generate common symbols for uninitialized data.
.IP "\fB\-fleading\-underscore\fR" 4
.IX Item "-fleading-underscore"
Add a leading underscore at the beginning of each C symbol.
.PP
Warning options:
.IP "\fB\-w\fR" 4
.IX Item "-w"
Disable all warnings.
.PP
Note: each of the following warning options has a negative form beginning with
\&\fB\-Wno\-\fR.
.IP "\fB\-Wimplicit\-function\-declaration\fR" 4
.IX Item "-Wimplicit-function-declaration"
Warn about implicit function declaration.
.IP "\fB\-Wunsupported\fR" 4
.IX Item "-Wunsupported"
Warn about unsupported \s-1GCC\s0 features that are ignored by \s-1TCC\s0.
.IP "\fB\-Wwrite\-strings\fR" 4
.IX Item "-Wwrite-strings"
Make string constants be of type \f(CW\*(C`const char *\*(C'\fR instead of \f(CW\*(C`char
*\*(C'\fR.
.IP "\fB\-Werror\fR" 4
.IX Item "-Werror"
Abort compilation if warnings are issued.
.IP "\fB\-Wall\fR" 4
.IX Item "-Wall"
Activate all warnings, except \fB\-Werror\fR, \fB\-Wunusupported\fR and
\&\fB\-Wwrite\-strings\fR.
.PP
Linker options:
.IP "\fB\-Ldir\fR" 4
.IX Item "-Ldir"
Specify an additional static library path for the \fB\-l\fR option. The
default library paths are \fI/usr/local/lib\fR, \fI/usr/lib\fR and \fI/lib\fR.
.IP "\fB\-lxxx\fR" 4
.IX Item "-lxxx"
Link your program with dynamic library libxxx.so or static library
libxxx.a. The library is searched in the paths specified by the
\&\fB\-L\fR option.
.IP "\fB\-shared\fR" 4
.IX Item "-shared"
Generate a shared library instead of an executable (\fB\-o\fR option
must also be given).
.IP "\fB\-static\fR" 4
.IX Item "-static"
Generate a statically linked executable (default is a shared linked
executable) (\fB\-o\fR option must also be given).
.IP "\fB\-rdynamic\fR" 4
.IX Item "-rdynamic"
Export global symbols to the dynamic linker. It is useful when a library
opened with \f(CW\*(C`dlopen()\*(C'\fR needs to access executable symbols.
.IP "\fB\-r\fR" 4
.IX Item "-r"
Generate an object file combining all input files (\fB\-o\fR option must
also be given).
.IP "\fB\-Wl,\-Ttext,address\fR" 4
.IX Item "-Wl,-Ttext,address"
Set the start of the .text section to \fIaddress\fR.
.IP "\fB\-Wl,\-\-oformat,fmt\fR" 4
.IX Item "-Wl,--oformat,fmt"
Use \fIfmt\fR as output format. The supported output formats are:
.RS 4
.ie n .IP """elf32\-i386""" 4
.el .IP "\f(CWelf32\-i386\fR" 4
.IX Item "elf32-i386"
\&\s-1ELF\s0 output format (default)
.ie n .IP """binary""" 4
.el .IP "\f(CWbinary\fR" 4
.IX Item "binary"
Binary image (only for executable output)
.ie n .IP """coff""" 4

ChangeLog
Details comments about changes are included in the git commit messages. This
file contains a brief summary of important changes in the system from one
version to the next.

4.0	xx August 2012

This version includes changes from the "classic" SPITBOL created from the 1980's on through 3.8.2

Changes include 

-- use of mcc, a compiler based on Philippe Bellard's tcc, instead of gcc.

-- use of nasm -- the netwide assembler -- instead of gas.

-- preference for lower case in source files. for example, minimal source no longer all upper case.

-- change in Minimal specification to more closely resemble nasm lexical conventions, such as ';' for comments.

-- file renaming related to translators to start using more systematic naming scheme.



3.8.3	July 25, 2012
This is the last version of "classic" SPITBOL dating back to the 1980's.
The only recent changes relate to the Linux port.

3.8.1	June 2012
This version completed the Linux port started by Mark Emmer in 2009.


mcc/mcc.c
/*
 *  MCC - Tiny C Compiler
 * 
 *  Copyright (c) 2001-2004 Fabrice Bellard
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "libmcc.c"

void help(void)
{
    printf("mcc version " MCC_VERSION " - Tiny C Compiler - Copyright (C) 2001-2006 Fabrice Bellard\n"
           "usage: mcc [-v] [-c] [-o outfile] [-Bdir] [-bench] [-Idir] [-Dsym[=val]] [-Usym]\n"
           "           [-Wwarn] [-g] [-b] [-bt N] [-Ldir] [-llib] [-shared] [-soname name]\n"
           "           [-static] [infile1 infile2...] [-run infile args...]\n"
           "\n"
           "General options:\n"
           "  -v          display current version, increase verbosity\n"
           "  -c          compile only - generate an object file\n"
           "  -o outfile  set output filename\n"
           "  -Bdir       set mcc internal library path\n"
           "  -bench      output compilation statistics\n"
           "  -run        run compiled source\n"
           "  -fflag      set or reset (with 'no-' prefix) 'flag' (see man page)\n"
           "  -Wwarning   set or reset (with 'no-' prefix) 'warning' (see man page)\n"
           "  -w          disable all warnings\n"
           "Preprocessor options:\n"
           "  -E          preprocess only\n"
           "  -Idir       add include path 'dir'\n"
           "  -Dsym[=val] define 'sym' with value 'val'\n"
           "  -Usym       undefine 'sym'\n"
           "Linker options:\n"
           "  -Ldir       add library path 'dir'\n"
           "  -llib       link with dynamic or static library 'lib'\n"
           "  -shared     generate a shared library\n"
           "  -soname     set name for shared library to be used at runtime\n"
           "  -static     static linking\n"
           "  -rdynamic   export all global symbols to dynamic linker\n"
           "  -r          generate (relocatable) object file\n"
           "Debugger options:\n"
           "  -g          generate runtime debug info\n"
#ifdef CONFIG_MCC_BCHECK
           "  -b          compile with built-in memory and bounds checker (implies -g)\n"
#endif
#ifdef CONFIG_MCC_BACKTRACE
           "  -bt N       show N callers in stack traces\n"
#endif
           );
}

static char **files;
static int nb_files, nb_libraries;
static int multiple_files;
static int print_search_dirs;
static int output_type;
static int reloc_output;
static const char *outfile;
static int do_bench = 0;

#define MCC_OPTION_HAS_ARG 0x0001
#define MCC_OPTION_NOSEP   0x0002 /* cannot have space before option and arg */

typedef struct MCCOption {
    const char *name;
    uint16_t index;
    uint16_t flags;
} MCCOption;

enum {
    MCC_OPTION_HELP,
    MCC_OPTION_I,
    MCC_OPTION_D,
    MCC_OPTION_U,
    MCC_OPTION_L,
    MCC_OPTION_B,
    MCC_OPTION_l,
    MCC_OPTION_bench,
    MCC_OPTION_bt,
    MCC_OPTION_b,
    MCC_OPTION_g,
    MCC_OPTION_c,
    MCC_OPTION_static,
    MCC_OPTION_shared,
    MCC_OPTION_soname,
    MCC_OPTION_o,
    MCC_OPTION_r,
    MCC_OPTION_Wl,
    MCC_OPTION_W,
    MCC_OPTION_O,
    MCC_OPTION_m,
    MCC_OPTION_f,
    MCC_OPTION_nostdinc,
    MCC_OPTION_nostdlib,
    MCC_OPTION_print_search_dirs,
    MCC_OPTION_rdynamic,
    MCC_OPTION_run,
    MCC_OPTION_v,
    MCC_OPTION_w,
    MCC_OPTION_pipe,
    MCC_OPTION_E,
};

static const MCCOption mcc_options[] = {
    { "h", MCC_OPTION_HELP, 0 },
    { "?", MCC_OPTION_HELP, 0 },
    { "I", MCC_OPTION_I, MCC_OPTION_HAS_ARG },
    { "D", MCC_OPTION_D, MCC_OPTION_HAS_ARG },
    { "U", MCC_OPTION_U, MCC_OPTION_HAS_ARG },
    { "L", MCC_OPTION_L, MCC_OPTION_HAS_ARG },
    { "B", MCC_OPTION_B, MCC_OPTION_HAS_ARG },
    { "l", MCC_OPTION_l, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "bench", MCC_OPTION_bench, 0 },
    { "bt", MCC_OPTION_bt, MCC_OPTION_HAS_ARG },
#ifdef CONFIG_MCC_BCHECK
    { "b", MCC_OPTION_b, 0 },
#endif
    { "g", MCC_OPTION_g, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "c", MCC_OPTION_c, 0 },
    { "static", MCC_OPTION_static, 0 },
    { "shared", MCC_OPTION_shared, 0 },
    { "soname", MCC_OPTION_soname, MCC_OPTION_HAS_ARG },
    { "o", MCC_OPTION_o, MCC_OPTION_HAS_ARG },
    { "run", MCC_OPTION_run, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "rdynamic", MCC_OPTION_rdynamic, 0 },
    { "r", MCC_OPTION_r, 0 },
    { "Wl,", MCC_OPTION_Wl, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "W", MCC_OPTION_W, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "O", MCC_OPTION_O, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "m", MCC_OPTION_m, MCC_OPTION_HAS_ARG },
    { "f", MCC_OPTION_f, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "nostdinc", MCC_OPTION_nostdinc, 0 },
    { "nostdlib", MCC_OPTION_nostdlib, 0 },
    { "print-search-dirs", MCC_OPTION_print_search_dirs, 0 }, 
    { "v", MCC_OPTION_v, MCC_OPTION_HAS_ARG | MCC_OPTION_NOSEP },
    { "w", MCC_OPTION_w, 0 },
    { "pipe", MCC_OPTION_pipe, 0},
    { "E", MCC_OPTION_E, 0},
    { NULL },
};

static int64_t getclock_us(void)
{
#ifdef _WIN32
    struct _timeb tb;
    _ftime(&tb);
    return (tb.time * 1000LL + tb.millitm) * 1000LL;
#else
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec * 1000000LL + tv.tv_usec;
#endif
}

static int strstart(const char *str, const char *val, const char **ptr)
{
    const char *p, *q;
    p = str;
    q = val;
    while (*q != '\0') {
        if (*p != *q)
            return 0;
        p++;
        q++;
    }
    if (ptr)
        *ptr = p;
    return 1;
}

/* convert 'str' into an array of space separated strings */
static int expand_args(char ***pargv, const char *str)
{
    const char *s1;
    char **argv, *arg;
    int argc, len;

    argc = 0;
    argv = NULL;
    for(;;) {
        while (is_space(*str))
            str++;
        if (*str == '\0')
            break;
        s1 = str;
        while (*str != '\0' && !is_space(*str))
            str++;
        len = str - s1;
        arg = mcc_malloc(len + 1);
        memcpy(arg, s1, len);
        arg[len] = '\0';
        dynarray_add((void ***)&argv, &argc, arg);
    }
    *pargv = argv;
    return argc;
}

int parse_args(MCCState *s, int argc, char **argv)
{
    int optind;
    const MCCOption *popt;
    const char *optarg, *p1, *r1;
    char *r;

    optind = 0;
    while (optind < argc) {

        r = argv[optind++];
        if (r[0] != '-' || r[1] == '\0') {
            /* add a new file */
            dynarray_add((void ***)&files, &nb_files, r);
            if (!multiple_files) {
                optind--;
                /* argv[0] will be this file */
                break;
            }
        } else {
            /* find option in table (match only the first chars */
            popt = mcc_options;
            for(;;) {
                p1 = popt->name;
                if (p1 == NULL)
                    error("invalid option -- '%s'", r);
                r1 = r + 1;
                for(;;) {
                    if (*p1 == '\0')
                        goto option_found;
                    if (*r1 != *p1)
                        break;
                    p1++;
                    r1++;
                }
                popt++;
            }
        option_found:
            if (popt->flags & MCC_OPTION_HAS_ARG) {
                if (*r1 != '\0' || (popt->flags & MCC_OPTION_NOSEP)) {
                    optarg = r1;
                } else {
                    if (optind >= argc)
                        error("argument to '%s' is missing", r);
                    optarg = argv[optind++];
                }
            } else {
                if (*r1 != '\0')
                    return 0;
                optarg = NULL;
            }
                
            switch(popt->index) {
            case MCC_OPTION_HELP:
                return 0;

            case MCC_OPTION_I:
                if (mcc_add_include_path(s, optarg) < 0)
                    error("too many include paths");
                break;
            case MCC_OPTION_D:
                {
                    char *sym, *value;
                    sym = (char *)optarg;
                    value = strchr(sym, '=');
                    if (value) {
                        *value = '\0';
                        value++;
                    }
                    mcc_define_symbol(s, sym, value);
                }
                break;
            case MCC_OPTION_U:
                mcc_undefine_symbol(s, optarg);
                break;
            case MCC_OPTION_L:
                mcc_add_library_path(s, optarg);
                break;
            case MCC_OPTION_B:
                /* set mcc utilities path (mainly for mcc development) */
                mcc_set_lib_path(s, optarg);
                break;
            case MCC_OPTION_l:
                dynarray_add((void ***)&files, &nb_files, r);
                nb_libraries++;
                break;
            case MCC_OPTION_bench:
                do_bench = 1;
                break;
#ifdef CONFIG_MCC_BACKTRACE
            case MCC_OPTION_bt:
                num_callers = atoi(optarg);
                break;
#endif
#ifdef CONFIG_MCC_BCHECK
            case MCC_OPTION_b:
                s->do_bounds_check = 1;
                s->do_debug = 1;
                break;
#endif
            case MCC_OPTION_g:
                s->do_debug = 1;
                break;
            case MCC_OPTION_c:
                multiple_files = 1;
                output_type = MCC_OUTPUT_OBJ;
                break;
            case MCC_OPTION_static:
                s->static_link = 1;
                break;
            case MCC_OPTION_shared:
                output_type = MCC_OUTPUT_DLL;
                break;
            case MCC_OPTION_soname:
                s->soname = optarg; 
                break;
            case MCC_OPTION_o:
                multiple_files = 1;
                outfile = optarg;
                break;
            case MCC_OPTION_r:
                /* generate a .o merging several output files */
                reloc_output = 1;
                output_type = MCC_OUTPUT_OBJ;
                break;
            case MCC_OPTION_nostdinc:
                s->nostdinc = 1;
                break;
            case MCC_OPTION_nostdlib:
                s->nostdlib = 1;
                break;
            case MCC_OPTION_print_search_dirs:
                print_search_dirs = 1;
                break;
            case MCC_OPTION_run:
                {
                    int argc1;
                    char **argv1;
                    argc1 = expand_args(&argv1, optarg);
                    if (argc1 > 0) {
                        parse_args(s, argc1, argv1);
                    }
                    multiple_files = 0;
                    output_type = MCC_OUTPUT_MEMORY;
                }
                break;
            case MCC_OPTION_v:
                do {
                    if (0 == s->verbose++)
                        printf("mcc version %s\n", MCC_VERSION);
                } while (*optarg++ == 'v');
                break;
            case MCC_OPTION_f:
                if (mcc_set_flag(s, optarg, 1) < 0 && s->warn_unsupported)
                    goto unsupported_option;
                break;
            case MCC_OPTION_W:
                if (mcc_set_warning(s, optarg, 1) < 0 && 
                    s->warn_unsupported)
                    goto unsupported_option;
                break;
            case MCC_OPTION_w:
                s->warn_none = 1;
                break;
            case MCC_OPTION_rdynamic:
                s->rdynamic = 1;
                break;
            case MCC_OPTION_Wl:
                {
                    const char *p;
                    if (strstart(optarg, "-Ttext,", &p)) {
                        s->text_addr = strtoul(p, NULL, 16);
                        s->has_text_addr = 1;
                    } else if (strstart(optarg, "--oformat,", &p)) {
                        if (strstart(p, "elf32-", NULL)) {
                            s->output_format = MCC_OUTPUT_FORMAT_ELF;
                        } else if (!strcmp(p, "binary")) {
                            s->output_format = MCC_OUTPUT_FORMAT_BINARY;
                        } else
#ifdef MCC_TARGET_COFF
                        if (!strcmp(p, "coff")) {
                            s->output_format = MCC_OUTPUT_FORMAT_COFF;
                        } else
#endif
                        {
                            error("target %s not found", p);
                        }
                    } else {
                        error("unsupported linker option '%s'", optarg);
                    }
                }
                break;
            case MCC_OPTION_E:
                output_type = MCC_OUTPUT_PREPROCESS;
                break;
            default:
                if (s->warn_unsupported) {
                unsupported_option:
                    warning("unsupported option '%s'", r);
                }
                break;
            }
        }
    }
    return optind + 1;
}

int main(int argc, char **argv)
{
    int i;
    MCCState *s;
    int nb_objfiles, ret, optind;
    char objfilename[1024];
    int64_t start_time = 0;

    s = mcc_new();
#ifdef _WIN32
    mcc_set_lib_path_w32(s);
#endif
    output_type = MCC_OUTPUT_EXE;
    outfile = NULL;
    multiple_files = 1;
    files = NULL;
    nb_files = 0;
    nb_libraries = 0;
    reloc_output = 0;
    print_search_dirs = 0;
    ret = 0;

    optind = parse_args(s, argc - 1, argv + 1);
    if (print_search_dirs) {
        /* enough for Linux kernel */
        printf("install: %s/\n", s->mcc_lib_path);
        return 0;
    }
    if (optind == 0 || nb_files == 0) {
        if (optind && s->verbose)
            return 0;
        help();
        return 1;
    }

    nb_objfiles = nb_files - nb_libraries;

    /* if outfile provided without other options, we output an
       executable */
    if (outfile && output_type == MCC_OUTPUT_MEMORY)
        output_type = MCC_OUTPUT_EXE;

    /* check -c consistency : only single file handled. XXX: checks file type */
    if (output_type == MCC_OUTPUT_OBJ && !reloc_output) {
        /* accepts only a single input file */
        if (nb_objfiles != 1)
            error("cannot specify multiple files with -c");
        if (nb_libraries != 0)
            error("cannot specify libraries with -c");
    }
    

    if (output_type == MCC_OUTPUT_PREPROCESS) {
        if (!outfile) {
            s->outfile = stdout;
        } else {
            s->outfile = fopen(outfile, "w");
            if (!s->outfile)
                error("could not open '%s", outfile);
        }
    } else if (output_type != MCC_OUTPUT_MEMORY) {
        if (!outfile) {
            /* compute default outfile name */
            char *ext;
            const char *name = 
                strcmp(files[0], "-") == 0 ? "a" : mcc_basename(files[0]);
            pstrcpy(objfilename, sizeof(objfilename), name);
            ext = mcc_fileextension(objfilename);
#ifdef MCC_TARGET_PE
            if (output_type == MCC_OUTPUT_DLL)
                strcpy(ext, ".dll");
            else
            if (output_type == MCC_OUTPUT_EXE)
                strcpy(ext, ".exe");
            else
#endif
            if (output_type == MCC_OUTPUT_OBJ && !reloc_output && *ext)
                strcpy(ext, ".o");
            else
                pstrcpy(objfilename, sizeof(objfilename), "a.out");
            outfile = objfilename;
        }
    }

    if (do_bench) {
        start_time = getclock_us();
    }

    mcc_set_output_type(s, output_type);

    /* compile or add each files or library */
    for(i = 0; i < nb_files && ret == 0; i++) {
        const char *filename;

        filename = files[i];
        if (filename[0] == '-' && filename[1]) {
            if (mcc_add_library(s, filename + 2) < 0) {
                error_noabort("cannot find %s", filename);
                ret = 1;
            }
        } else {
            if (1 == s->verbose)
                printf("-> %s\n", filename);
            if (mcc_add_file(s, filename) < 0)
                ret = 1;
        }
    }

    /* free all files */
    mcc_free(files);

    if (ret)
        goto the_end;

    if (do_bench)
        mcc_print_stats(s, getclock_us() - start_time);

    if (s->output_type == MCC_OUTPUT_PREPROCESS) {
        if (outfile)
            fclose(s->outfile);
    } else if (s->output_type == MCC_OUTPUT_MEMORY) {
        ret = mcc_run(s, argc - optind, argv + optind);
    } else
        ret = mcc_output_file(s, outfile) ? 1 : 0;
 the_end:
    /* XXX: cannot do it with bound checking because of the malloc hooks */
    if (!s->do_bounds_check)
        mcc_delete(s);

#ifdef MEM_DEBUG
    if (do_bench) {
        printf("memory: %d bytes, max = %d bytes\n", mem_cur_size, mem_max_size);
    }
#endif
    return ret;
}

